### PR TITLE
ipn/ipnserver, log/filelogger, logpolicy: when tailscaled is running …

### PR DESCRIPTION
**Pull request by dblohm7**
…as a windows service, ensure the service's log messages are written to file

This patch moves the Windows-only initialization of the filelogger into logpolicy.
Previously we only did it when babysitting the tailscaled subprocess, but this meant
that log messages from the service itself never made it to disk. This meant that
if logtail could not dial out, its log messages would be lost.

I modified filelogger.New to work a bit differently and added a `maybeWrapForPlatform`
to logpolicy to ensure that the filelogger is plugged in front of logtail ASAP.

Fixes https://github.com/tailscale/tailscale/issues/3570

Signed-off-by: Aaron Klotz <aaron@tailscale.com>

### CODE REVIEW

**Comment by bradfitz** [comment]
Let's chat about all this when you're around.

**Comment by bradfitz** [comment]
> log messages from the service itself never made it to disk.

Which log messages are those?

> This meant that if logtail could not dial out, its log messages would be lost.

It's intentional that logtail's errors (as used by the logwriter backoff `bo` field) never loop back into that log writer, otherwise it'd amplify. This is purely about those logtail dial error messages not going to disk?

**Comment by dblohm7** [comment]
> Which log messages are those?

Any log message generated by the parent service.

> It's intentional that logtail's errors (as used by the logwriter backoff `bo` field) never loop back into that log writer, otherwise it'd amplify.

That makes sense. I do find it hard to get my head around all the complexity w/ respect to who is plugged into what.

> This is purely about those logtail dial error messages not going to disk?

That would be the main impetus, yes, though there are some other useful entries that wouldn't make it to disk (such as DNS cache flushing upon screen unlock, when enabled)

**Comment by dblohm7** [comment]
Though I don't think that's a concern here, as the filelogger is essentially a tee off the existing setup.

File: log/filelogger/log.go [hunk 1]
```diff
@@ -26,30 +27,30 @@ const (
 	maxFiles = 50
 )
 
-// New returns a logf wrapper that appends to local disk log
+// New returns a Writer that appends to local disk log
 // files on Windows, rotating old log files as needed to stay under
 // file count & byte limits.
-func New(fileBasePrefix, logID string, logf logger.Logf) logger.Logf {
+func New(fileBasePrefix, logID string, inner *log.Logger) io.Writer {
```
Comment (bradfitz) [inline log/filelogger/log.go:33]: so, I like everything about this PR (like moving the file logging up from ipnserver) except this signature change. An io.Writer is logically a byte stream and not record oriented like logf is. Generally when an io.Writer is used record-wise (or a passed in one has that constraint), it's documented, like the `log` package saying:

> Each logging operation makes a single call to the Writer's Write method.

That'd be one option, to just document it here.

But I'd prefer to stay in the `logf` type world if possible.

I see why that's tricky, though... logpolicy is what generates the LogID.

Still thinking.

Comment (bradfitz) [reply log/filelogger/log.go:33]: What do you think about #3582?

**Pull request closed without merging**